src/c++/perf_analyzer/docs/llm.md

```diff
@@ -0,0 +1,136 @@
+            ],
+            "STREAM": [
+                true
+            ]
+        }
+    ]
+}
+' > prompts.json
+```
+
+#### 3. Run PA
+
+```bash
+perf_analyzer \
+    -m vllm \
+    -i grpc \
+    --async \
+    --streaming \
+    --input-data=prompts.json \
+    --profile-export-file=profile_export.json \
+    --measurement-mode=count_windows \
+    --measurement-request-count=10 \
+    --stability-percentage=999
+```
+
+#### 4. Calculate average time to first token
+
+```bash
+python3 -c 'import json
+f = open("profile_export.json")
+requests = json.load(f)["experiments"][0]["requests"]
+latencies = [r["response_timestamps"][0] - r["timestamp"] for r in requests]
+avg_latency_s = sum(latencies) / len(latencies) / 1000000000
+print("Average time to first token: " + str(avg_latency_s) + " s")'
+
+# Average time to first token: 0.672695455483871 s
+```
+
+#### 5. Repeat steps 3-4 with different prompt lengths to measure effects of initial prompt size (prefill) on first-token latency.
+
+### Generation
+
+In this benchmarking scenario, we want to measure the effect of input prompt size on token-to-token latency. We issue single request to the server of fixed input sizes and request the model to compute a fixed amount of tokens.
+
+#### 1. Run these commands
+
+```bash
+PATH_TO_MODEL_PY="model_repository/vllm/1/model.py"
+MAX_TOKENS=256
+sed -i "128s/.*/\ \ \ \ \ \ \ \ params_dict[\"max_tokens\"] = ${MAX_TOKENS}/" ${PATH_TO_MODEL_PY}
+```
+
+#### 2. Follow step 2 from the [Triton + vLLM tutorial](https://github.com/triton-inference-server/tutorials/blob/main/Quick_Deploy/vLLM/README.md).
+
+#### 3. Generate prompts input data JSON
+
+```bash
+echo '
+{
+    "data": [
+        {
+            "PROMPT": [
+                "Hello, my name is"
+            ],
+            "STREAM": [
+                true
+            ]
+        }
+    ]
+}
+' > prompts.json
+```
+
+#### 3. Run PA
+
+```bash
+perf_analyzer \
+    -m vllm \
+    -i grpc \
+    --async \
+    --streaming \
+    --input-data=prompts.json \
+    --profile-export-file=profile_export.json \
+    --measurement-mode=count_windows \
+    --measurement-request-count=10 \
+    --stability-percentage=999
+```
+
+#### 4. Calculate average time to first token
+
+```bash
+python3 -c 'import json
+f = open("/workspace/tmp/profile_export.json")
+requests = json.load(f)["experiments"][0]["requests"]
+latencies = []
+for request in requests:
+    prev_response = request["response_timestamps"][0]
+    for response in request["response_timestamps"][1:]:
+        latencies.append(response - prev_response)
+        prev_response = response
+avg_latency_s = sum(latencies) / len(latencies) / 1000000000
+print("Average token-to-token latency: " + str(avg_latency_s) + " s")'
```

Comment:
Is it possible to make this a separate Python script (e.g., `calculate_avg_token_to_token_latency.py`) and just show how to run this script in here?

Reply:
Done 👍